### PR TITLE
Add configurable exit grace periods and shutdown handlers

### PR DESCRIPTION
**Pull request by kpayson64**

### CODE REVIEW

File: src/python/grpcio/grpc/_server.py [hunk 1]
```diff
@@ -719,22 +727,26 @@ def _start(state):
     _request_call(state)    
     def cleanup_server(timeout):
       if timeout is None:
-        _stop(state, _UNEXPECTED_EXIT_SERVER_GRACE).wait()
+        _stop(state, state.exit_grace, state.exit_handler_grace).wait()
       else:
-        _stop(state, timeout).wait()
+        _stop(state, timeout, True).wait()
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_server.py]: I don't understand the `True` on this line. Isn't a numeric value or `None` appropriate in that position?

File: src/python/grpcio/grpc/_server.py [hunk 2]
```diff
@@ -719,22 +727,26 @@ def _start(state):
...
-  def __init__(self, thread_pool, generic_handlers, options):
+  def __init__(self, thread_pool, generic_handlers, options, exit_grace,
+               exit_handler_grace):
     completion_queue = cygrpc.CompletionQueue()
     server = cygrpc.Server(_common.channel_args(options))
     server.register_completion_queue(completion_queue)
     self._state = _ServerState(
-        completion_queue, server, generic_handlers, thread_pool)
+        completion_queue, server, generic_handlers, thread_pool,
+        exit_grace if exit_grace is not None else _DEFAULT_EXIT_GRACE,
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_server.py]: Prefer avoiding `not` when you can; flip these branches and drop the `not`.

File: src/python/grpcio/grpc/_server.py [hunk 3]
```diff
@@ -719,22 +727,26 @@ def _start(state):
...
+        exit_handler_grace if exit_handler_grace is not None else
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_server.py]: Here too.

File: src/python/grpcio/grpc/_server.py (hunk 1 above)
Comment (kpayson64) [inline src/python/grpcio/grpc/_server.py]: Oops, good catch.  I changed the API when I was working on this and this is a vestige.

**Comment by kpayson64** [comment]
Addressed comments

File: src/python/grpcio/grpc/__init__.py [hunk 4]
```diff
@@ -905,6 +905,21 @@ def add_secure_port(self, address, server_credentials):
     raise NotImplementedError()
 
   @abc.abstractmethod
+  def add_shutdown_handler(self, shutdown_handler):
+    """Adds a shutdown handler to be called on server shutdown.
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/__init__.py]: Drop the first "handler" in this sentence - the term "shutdown handler" isn't defined until the reader reads the rest of the sentence and infers that being called on server shutdown is what makes a handler a "shutdown handler".

File: src/python/grpcio/grpc/__init__.py [hunk 5]
```diff
@@ -905,6 +905,21 @@ def add_secure_port(self, address, server_credentials):
...
+
+    Shutdown handlers are run on server stop() or in the event that a running
+    server is destroyed unexpectedly.  The handlers are run in series before
+    the stop grace period.
+
+    Args:
+      shutdown_handler:  A function taking a single arg, a time in seconds that
+      the handler should complete within.  None indicates that that handler can
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/__init__.py]: Nit: "within which the handler should complete".

Okay, okay: "pico-nit".

File: src/python/grpcio/grpc/_server.py [hunk 6]
```diff
@@ -672,41 +677,44 @@ def _serve(state):
...
-      shutdown_event.set()
-      return shutdown_event
-    else:
-      if state.stage is _ServerStage.STARTED:
-        state.server.shutdown(state.completion_queue, _SHUTDOWN_TAG)
+def _stop(state, grace, shutdown_handler_grace):
+  shutdown_event = threading.Event()
+
+  def cancel_all_calls_after_grace():
+    do_shutdown = False
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_server.py]: Please favor
```
if <condition>:
  <name> = <expression>
else:
  <name> = <other expression>
```
over
```
<name> = <other expression>
if <condition>:
  <name> = <expression>
```
. Here, the general form particularizes to
```
<*no* assignment of do_shutdown>
with state.lock:
  if <one condition>:
    <statements that return>
  elif <another condition>:
    <statements that don't return>
    do_shutdown = True
  else:
    do_shutdown = False
```
.

**Comment by kpayson64** [comment]
Addressed comments.  The untested code path is almost impossible to test.

The join() function that exposes it is only called by the internals of threading.py, and we don't hold a reference to the server thread to explicitly join() it, and I'm not sure we should add a reference just for this purpose.

Moreover, the threading.py only calls join(None), the code path in question isn't even exercised by the internals of threading.py.  Its just there to make sure we properly follow the join(timeout) semantics.

**Comment by nathanielmanistaatgoogle** [comment]
Change content looks good, but the detailed response you just gave me about the difficulty of testing some parts of this change belongs in the commit message.

**Comment by kpayson64** [comment]
Addressed comment.

**Comment by nathanielmanistaatgoogle** [comment]
In your commit log message I think you mean "It's just there" rather than "Its just there".

**Comment by kpayson64** [comment]
@soltanmm-google ping

File: src/python/grpcio/grpc/_server.py [hunk 7]
```diff
@@ -745,11 +758,14 @@ def add_insecure_port(self, address):
...
 
+  def add_shutdown_handler(self, handler):
+    self._state.shutdown_handlers.append(handler)
+
   def start(self):
     _start(self._state)
 
-  def stop(self, grace):
-    return _stop(self._state, grace)
+  def stop(self, grace, shutdown_handler_grace=None):
```
Comment (soltanmm-google) [inline src/python/grpcio/grpc/_server.py]: nit-question: any particular reason why this is called `shutdown_handler_grace` and other places call it `exit_handler_grace`?

Comment (kpayson64) [inline src/python/grpcio/grpc/_server.py]: ```shutdown_handler_grace``` and ```exit_handler_grace``` are different things.  ```shutdown_handler_grace``` is the grace period passed to to the shutdown handlers, ```exit_handler_grace``` is the ```shutdown_handler_grace``` to be used in the event of a server running on exit.  I've renamed ```exit_handler_grace``` ->```exit_shutdown_handler_grace``` to make that more clear.

**Comment by kpayson64** [comment]
Addressed comments, back to @soltanmm-google

**Comment by soltanmm-google** [comment]
lgtm

**Comment by nathanielmanistaatgoogle** [comment]
(Squash before merging, please.)

**Pull request merged**